### ISSUE TITLE
Continual training

### ISSUE THREAD

**Issue by dk-teknologisk-mlnn** [ultralytics/ultralytics#10118]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I trained a detection model in yolov5 and it has around  95% prec recall and map50. 
I trained it for 3000 epochs in yolov8 and got ~85%. I was like " hmm, maybe just another 1000 epochs then!"
So I found questions about this continuous training and how to avoid that the first epoch destroy the model performance.  and they said "make sure to have 0 warmup epochs". 
So I did and I set initial LR to be the end LR of the 3000 epoch model. but still the performance is reduced right of the bat, see image. orange is the first training and red is hte one that tries to continue from the orange model.
![continualtraining](https://github.com/ultralytics/ultralytics/assets/30144796/d5a7ed46-7cc7-4bc2-a69e-b10edcf11b31)

Is there any way to keep the initial performance of a loaded pretrained model intact?

**Comment by glenn-jocher** [comment]
Hi there! 😀 Thank you for bringing up your question regarding continual training with YOLOv8 models.

Continuing training from a pretrained model can sometimes cause an initial performance drop, as you've observed. It seems like you've already taken a great step by setting the warmup epochs to 0 and adjusting the initial learning rate (LR) to match the end LR from your previous training session. That's the right direction!

Here's a small additional step you might consider to help maintain the initial performance:
- **Experiment with a Lower Learning Rate:** Even if you start with the final LR from your last training session, further reducing the initial LR slightly for the continuation can help prevent abrupt performance drops. Sometimes, very small adjustments are beneficial.

For example:
```bash
yolo detect train model=your_model.pt epochs=1000 lr0=0.0001
```
(Note: Adjust `lr0` based on your end LR from the 3000 epochs.)

Also, ensure your dataset and training configurations haven't unintentionally changed. Small inconsistencies can also impact performance.

If these steps don't maintain the initial performance, consider evaluating if the additional training epochs are introducing overfitting.

Hope this helps! If you encounter any other issues or have further questions, feel free to ask.

**Comment by dk-teknologisk-mlnn** [comment]
Isnt it also because the optimizer from the old training is not stored and restored from the model?
in another project I store checkpoints with the model state dicts, epoch, lr_scheduler and optimizer state dicts as well. that way I can load them in the new training and the history for adaptive learning and momentum is kept.

**Comment by glenn-jocher** [comment]
@dk-teknologisk-mlnn glad to hear that adjusting the LR helped with SGD! 😊 About using Adam and other adaptive optimizers, it's indeed because the optimizer state isn't retained between training sessions. Saving optimizer states, along with the model state dict, LR scheduler, and epoch info (just like in your other project), is a great approach to preserve the training context. This practice can lead to a smoother continuation of training, as it keeps the optimizer's momentum and adaptive learning history intact.

Here's a quick snippet on how you might save these components:

```python
# Example of saving model, optimizer state, and epoch info
torch.save({
    'model_state_dict': model.state_dict(),
    'optimizer_state_dict': optimizer.state_dict(),
    'lr_scheduler': scheduler.state_dict(),
    'epoch': epoch,
}, 'checkpoint.pt')
```

And loading them back:

```python
# Assuming model, optimizer, and scheduler have been defined
checkpoint = torch.load('checkpoint.pt')
model.load_state_dict(checkpoint['model_state_dict'])
optimizer.load_state_dict(checkpoint['optimizer_state_dict'])
scheduler.load_state_dict(checkpoint['lr_scheduler'])
current_epoch = checkpoint['epoch']
```

Feel free to experiment with this approach and see how it impacts your training with Adam or other optimizers. 😄
